lib/pdf/core/page.rb

```diff
@@ -22,18 +22,27 @@ def initialize(document, options = {})
         end
       end
 
+      # If :Contents is a reference to an array, returns the resolved reference.
+      # Otherwise, makes sure :Contents is an array.
+      def ensure_contents_array
+        contents = dictionary.data[:Contents]
+        if contents.is_a?(PDF::Core::Reference) && contents.data.is_a?(Array)
+          return contents.data
+        end
+        # Ensure contents is an array.
+        dictionary.data[:Contents] = Array(contents)
+      end
+
       # As per the PDF spec, each page can have multiple content streams. This
       # will add a fresh, empty content stream this the page, mainly for use in
       # loading template files.
       #
       def new_content_stream
         return if in_stamp_stream?
 
-        unless dictionary.data[:Contents].is_a?(Array)
-          dictionary.data[:Contents] = [content]
-        end
         @content = document.ref({})
-        dictionary.data[:Contents] << document.state.store[@content]
+        contents = ensure_contents_array
```

Comment:
Would it be possible to split this in two distinct operations? `ensure_contents_array` looks a lot like a command and return value here appears to be circumstantial.